indexer/indexer.py

```diff
@@ -0,0 +1,156 @@
+                            time.sleep(self.rejected_publish_delay)
+
+            except pika.exceptions.AMQPConnectionError as exception:
+                self.log.error(f"AMQP Connection Error: {exception}")
+                self.log.info("Reconnecting to RabbitMQ and database")
+                self.on_error(exception)
+            except exc.DBAPIError as exception:
+                self.log.error(f"Database connection error: {exception}")
+                self.log.info("Reconnecting to RabbitMQ and database")
+                self.on_error(exception)
+            except KeyboardInterrupt:
+                self.log.info(" Exiting...")
+                try:
+                    rmq_connection.close()
+                except NameError:
+                    pass
+                sys.exit(0)
+
+    def _connect_to_rabbitmq(self):
+        while True:
+            try:
+                self.log.info(f"Connecting to RabbitMQ ({self.rabbitmq_parameters.host}:"
+                              f"{self.rabbitmq_parameters.port})...")
+                rmq_connection = pika.BlockingConnection(pika.ConnectionParameters(host=self.rabbitmq_parameters.host,
+                                                                                   port=self.rabbitmq_parameters.port))
+                channel = rmq_connection.channel()
+                self.log.info("Connected to RabbitMQ")
+                channel.confirm_delivery()
+                channel.queue_declare(queue=self.QUEUE_NAME, durable=True)
+                return rmq_connection, channel
+            except pika.exceptions.AMQPConnectionError as exception:
+                self.log.error(f"AMQP Connection Error: {exception}")
+                time.sleep(5)
+
+    def _connect_to_database(self, database: Database):
+        while True:
+            try:
+                self.log.info("Connecting to a database")
+                database.connect()
+                self.log.info("Connected to a database")
+                return
+            except exc.DBAPIError as exception:
+                self.log.error(f"Database connection error: {exception}")
+                time.sleep(5)
+                if not exception.connection_invalidated:
+                    raise exception
+
+    def _prepare_downloader_message(self, crawl_result: CrawlResult) -> bytes:
+        message = {
+            "repo_id": self._create_repo_id(crawl_result.id),
+            "git_url": crawl_result.git_url
+        }
+        if crawl_result.languages is not None:
+            message['languages'] = [language.value for language in crawl_result.languages.keys()]
+
+        return bytes(json.dumps(message), encoding='utf-8')
+
+    def _create_repo_id(self, crawled_id: str):
+        return f"{self.indexer_type.value['id_prefix']}_{crawled_id}"
+
+    @abstractmethod
+    def crawl_next_repository(self, prev_repository_id: Optional[str]) -> CrawlResult:
```

Comment:
I thinking about replacing this method by a generator but I'm not sure if it will be simple to understand for everyone